parser2/src/tokenize.rs

```diff
@@ -0,0 +1,585 @@
+//     }
+// }
+
+// pub type Spanned<T> = (Location, T, Location);
+pub type Spanned<T> = (usize, T, usize);
+
+pub type SpannedToken<'input> = Spanned<Token<'input>>;
+
+// #[derive(Copy, Clone, PartialEq, Debug)]
+// enum Delimiter {
+//     Brace,
+//     Bracket,
+//     Paren,
+// }
+
+// #[derive(Clone, Debug, PartialEq)]
+// enum Context {
+//     Block {
+//         emit_semi: bool,
+//         needs_close: bool,
+//     },
+//     Expr,
+//     Let,
+//     Type,
+//     If,
+//     Then,
+//     Brace,
+//     Bracket,
+//     Paren,
+//     MatchArm,
+//     Lambda,
+// }
+
+// #[derive(Clone, Debug)]
+// struct Offside {
+//     context: Context,
+//     column: usize,
+// }
+
+// struct Contexts {
+//     stack: Vec<Offside>,
+// }
+
+// impl Contexts {
+//     fn new() -> Contexts {
+//         Contexts {
+//             stack: vec![],
+//         }
+//     }
+
+//     fn last(&self) -> Option<&Offside> {
+//         self.stack.last()
+//     }
+
+//     fn last_mut(&mut self) -> Option<&mut Offside> {
+//         self.stack.last_mut()
+//     }
+
+//     fn pop(&mut self) -> Option<Offside> {
+//         self.stack.pop()
+//     }
+
+//     fn push(&mut self, offside: Offside) -> Result<(), Error> {
+//         try!(self.check_unindentation_limit(&offside));
+//         self.stack.push(offside);
+//         Ok(())
+//     }
+
+//     fn check_unindentation_limit(&mut self, offside: &Offside) -> Result<(), Error> {
+//         let mut skip_block = false;
+
+//         for other_offside in self.stack.iter().rev() {
+//             match other_offside.context {
+//                 Context::Lambda => skip_block = true,
+//                 Context::Brace |
+//                 Context::Bracket |
+//                 Context::Paren => return Ok(()),
+//                 Context::Block { .. } if skip_block => (),
+
+//                 // New context should not be unindented past the closest enclosing block context
+//                 Context::MatchArm |
+//                 Context::Type |
+//                 Context::Let |
+//                 Context::Block { .. } if offside.column < other_offside.column => {
+//                     return Err(Error::UnderIndented)
+//                 }
+
+//                 Context::MatchArm |
+//                 Context::Type |
+//                 Context::Let |
+//                 Context::Block { .. } |
+//                 Context::Expr |
+//                 Context::If |
+//                 Context::Then => (),
+//             }
+//         }
+
+//         Ok(())
+//     }
+// }
+
+pub struct Tokenizer<'input> {
+    src: &'input str,
+    chars: CharIndices<'input>,
+    lookahead: Option<(usize, char)>,
+    // indent_levels: Contexts,
+}
+
+fn is_operator_symbol(ch: char) -> bool {
+    "+-*/&|=<>:.".chars().any(|x| x == ch)
+}
+
+fn is_ident_start(ch: char) -> bool {
+    UnicodeXID::is_xid_start(ch)
+}
+
+fn is_ident_continue(ch: char) -> bool {
+    UnicodeXID::is_xid_continue(ch)
```

Comment:
Wasn't aware that unicode had specifications for identifiers, is this commonly used in other languages?